Abstract view annotated with @Route gets published

If you create an abstract super class for your views, that gets published as a route, but naturally fails when user tries to navigate there.

Tested with 12.0.6, with following code snippet:

```
@Route
public abstract class AbstractView extends VerticalLayout {
    
}

public class SomeView extends AbstractView {

    public SomeView() {
        add(new Paragraph("Some"));
    }

}
```

In Spring apps this currently works correctly, in CDI apps only the abstract class gets published https://github.com/vaadin/cdi/issues/279 .

I'm not sure about `gets published as a route` meaning.

In my case when I run a test I'm getting the exception:
```
java.lang.RuntimeException: javax.servlet.ServletException: Exception while registering Routes on servlet startup
    at org.eclipse.jetty.annotations.ServletContainerInitializersStarter.doStart (ServletContainerInitializersStarter.java:69)
```
with a cause :

```
Caused by: com.vaadin.flow.server.InvalidRouteConfigurationException: Navigation targets must have unique routes, found navigation targets 'com.vaadin.flow.RealRoute' and 'com.vaadin.flow.AbstractRoute' with the same route.
    at com.vaadin.flow.server.startup.RouteTarget.validateNormalTarget
```

So we should improve our validation at startup  and don't register a route whose class is an interface or an abstract class.

--- a/flow-server/src/main/java/com/vaadin/flow/router/internal/AbstractRouteRegistry.java
+++ b/flow-server/src/main/java/com/vaadin/flow/router/internal/AbstractRouteRegistry.java
@@ -56,7 +56,7 @@ public abstract class AbstractRouteRegistry implements RouteRegistry {
          * Configure the given configurable route holder object.
          *
          * @param configuration
-         *         mutable route configuration to make changes to
+         *            mutable route configuration to make changes to
          */
         void configure(ConfigureRoutes configuration);
     }
@@ -81,7 +81,7 @@ public abstract class AbstractRouteRegistry implements RouteRegistry {
      * Thread-safe update of the RouteConfiguration.
      *
      * @param command
-     *         command that will mutate the configuration copy.
+     *            command that will mutate the configuration copy.
      */
     protected void configure(Configuration command) {
         lock();
@@ -148,7 +148,7 @@ public abstract class AbstractRouteRegistry implements RouteRegistry {
      * Fire routes changed event to all registered listeners.
      *
      * @param routeChangedEvent
-     *         event containing changes
+     *            event containing changes
      */
     protected void fireEvent(RoutesChangedEvent routeChangedEvent) {
         routesChangedListeners
@@ -176,12 +176,15 @@ public abstract class AbstractRouteRegistry implements RouteRegistry {
      * @return current state of the registry as a value object
      */
     public ConfiguredRoutes getConfiguration() {
+        if (configurationLock.isHeldByCurrentThread() && editing != null) {
+            return editing;
+        }
         return configuredRoutes;
     }
 
     @Override
     public List<RouteData> getRegisteredRoutes() {
-        return getRegisteredRoutes(configuredRoutes);
+        return getRegisteredRoutes(getConfiguration());
     }
 
     private List<RouteData> getRegisteredRoutes(
@@ -193,8 +196,8 @@ public abstract class AbstractRouteRegistry implements RouteRegistry {
             List<RouteAliasData> routeAliases = new ArrayList<>();
 
             configuration.getRoutePaths(target).stream()
-                    .filter(route -> !route.equals(url)).forEach(
-                    route -> routeAliases.add(new RouteAliasData(
+                    .filter(route -> !route.equals(url))
+                    .forEach(route -> routeAliases.add(new RouteAliasData(
                             getParentLayouts(configuration, target, route),
                             route, parameters, target)));
             List<Class<? extends RouterLayout>> parentLayouts = getParentLayouts(
@@ -215,7 +218,7 @@ public abstract class AbstractRouteRegistry implements RouteRegistry {
      * even if aliases change.
      *
      * @param routeData
-     *         route data to flatten.
+     *            route data to flatten.
      * @return flattened list of routes and aliases
      */
     private List<RouteBaseData<?>> flattenRoutes(List<RouteData> routeData) {
@@ -275,23 +278,22 @@ public abstract class AbstractRouteRegistry implements RouteRegistry {
      * Append any required parameters as /{param_class} to the route.
      *
      * @param navigationTarget
-     *         navigation target to generate url for
+     *            navigation target to generate url for
      * @return route with required parameters
      */
     private String collectRequiredParameters(
             Class<? extends Component> navigationTarget) {
-        if (!configuredRoutes.hasRouteTarget(navigationTarget)) {
+        if (!getConfiguration().hasRouteTarget(navigationTarget)) {
             return null;
         }
         StringBuilder route = new StringBuilder(
-                configuredRoutes.getTargetRoute(navigationTarget));
+                getConfiguration().getTargetRoute(navigationTarget));
 
         List<Class<?>> routeParameters = getRouteParameters(navigationTarget);
 
         if (!routeParameters.isEmpty()) {
-            routeParameters.forEach(
-                    param -> route.append("/{").append(param.getSimpleName())
-                            .append("}"));
+            routeParameters.forEach(param -> route.append("/{")
+                    .append(param.getSimpleName()).append("}"));
         }
         return route.toString();
     }
@@ -329,8 +331,8 @@ public abstract class AbstractRouteRegistry implements RouteRegistry {
         if (!getConfiguration().hasRoute(path)) {
             return;
         }
-        configure(configuration -> configuration
-                .removeRoute(path, navigationTarget));
+        configure(configuration -> configuration.removeRoute(path,
+                navigationTarget));
     }
 
     @Override
@@ -344,11 +346,11 @@ public abstract class AbstractRouteRegistry implements RouteRegistry {
      * Note! this should only be called from a configure() for thread safety.
      *
      * @param path
-     *         path for the navigation target
+     *            path for the navigation target
      * @param navigationTarget
-     *         navigation target for given path
+     *            navigation target for given path
      * @param configuration
-     *         mutable configuration object
+     *            mutable configuration object
      * @return the route target to which the target was added
      */
     private RouteTarget addRouteToConfiguration(String path,
@@ -374,12 +376,12 @@ public abstract class AbstractRouteRegistry implements RouteRegistry {
      * related.
      *
      * @param target
-     *         error handler target
+     *            error handler target
      * @param exceptionTargetsMap
-     *         map of existing error handlers
+     *            map of existing error handlers
      * @throws InvalidRouteConfigurationException
-     *         if trying to add a non related exception handler for which a
-     *         handler already exists
+     *             if trying to add a non related exception handler for which a
+     *             handler already exists
      */
     protected void addErrorTarget(Class<? extends Component> target,
             Map<Class<? extends Exception>, Class<? extends Component>> exceptionTargetsMap) {
@@ -404,9 +406,9 @@ public abstract class AbstractRouteRegistry implements RouteRegistry {
      * allowed.
      *
      * @param target
-     *         target being handled
+     *            target being handled
      * @param exceptionType
-     *         type of the handled exception
+     *            type of the handled exception
      */
     private void handleRegisteredExceptionType(
             Map<Class<? extends Exception>, Class<? extends Component>> exceptionTargetsMap,
@@ -418,10 +420,10 @@ public abstract class AbstractRouteRegistry implements RouteRegistry {
         if (registered.isAssignableFrom(target)) {
             exceptionTargetsMap.put(exceptionType, target);
         } else if (!target.isAssignableFrom(registered)) {
-            String msg = String
-                    .format("Only one target for an exception should be defined. Found '%s' and '%s' for exception '%s'",
-                            target.getName(), registered.getName(),
-                            exceptionType.getName());
+            String msg = String.format(
+                    "Only one target for an exception should be defined. Found '%s' and '%s' for exception '%s'",
+                    target.getName(), registered.getName(),
+                    exceptionType.getName());
             throw new InvalidRouteLayoutConfigurationException(msg);
         }
     }
@@ -431,16 +433,16 @@ public abstract class AbstractRouteRegistry implements RouteRegistry {
      * cause until possible exception with handler found.
      *
      * @param exception
-     *         exception to get handler for
+     *            exception to get handler for
      * @return Optional containing found handler or empty if none found
      */
     protected Optional<ErrorTargetEntry> searchByCause(Exception exception) {
-        Class<? extends Component> targetClass = configuredRoutes
+        Class<? extends Component> targetClass = getConfiguration()
                 .getExceptionHandlerByClass(exception.getClass());
 
         if (targetClass != null) {
-            return Optional.of(new ErrorTargetEntry(targetClass,
-                    exception.getClass()));
+            return Optional.of(
+                    new ErrorTargetEntry(targetClass, exception.getClass()));
         }
 
         Throwable cause = exception.getCause();
@@ -455,15 +457,15 @@ public abstract class AbstractRouteRegistry implements RouteRegistry {
      * exist.
      *
      * @param exception
-     *         exception to get handler for
+     *            exception to get handler for
      * @return Optional containing found handler or empty if none found
      */
     protected Optional<ErrorTargetEntry> searchBySuperType(
             Throwable exception) {
         Class<?> superClass = exception.getClass().getSuperclass();
-        while (superClass != null && Exception.class
-                .isAssignableFrom(superClass)) {
-            Class<? extends Component> targetClass = configuredRoutes
+        while (superClass != null
+                && Exception.class.isAssignableFrom(superClass)) {
+            Class<? extends Component> targetClass = getConfiguration()
                     .getExceptionHandlerByClass(superClass);
             if (targetClass != null) {
                 return Optional.of(new ErrorTargetEntry(targetClass,

--- a/flow-server/src/main/java/com/vaadin/flow/server/AmbiguousRouteConfigurationException.java
+++ b/flow-server/src/main/java/com/vaadin/flow/server/AmbiguousRouteConfigurationException.java
@@ -1,0 +1,66 @@
+/*
+ * Copyright 2000-2018 Vaadin Ltd.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License"); you may not
+ * use this file except in compliance with the License. You may obtain a copy of
+ * the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
+ * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
+ * License for the specific language governing permissions and limitations under
+ * the License.
+ */
+package com.vaadin.flow.server;
+
+import com.vaadin.flow.component.Component;
+
+/**
+ *
+ * Exception indicating that the application's routes already has the navigation
+ * target with the given path.
+ *
+ * @author Vaadin Ltd
+ *
+ */
+public class AmbiguousRouteConfigurationException
+        extends InvalidRouteConfigurationException {
+
+    private final Class<? extends Component> navigationTarget;
+
+    /**
+     * Constructs a new invalid route configuration exception with the specified
+     * detail message and the existing navigation target component which already
+     * presents in the configuration with the route path.
+     *
+     * @param message
+     *            the detail message. The detail message is saved for later
+     *            retrieval by the {@link #getMessage()} method.
+     * @param navigationTarget
+     *            the configured navigation target, not {@code null}
+     */
+    public AmbiguousRouteConfigurationException(String message,
+            Class<? extends Component> navigationTarget) {
+        super(message);
+        this.navigationTarget = navigationTarget;
+    }
+
+    /**
+     * Returns the already configured navigation target component class which
+     * caused the exception.
+     * <p>
+     * In case the exception happens as a result of a navigation target
+     * collision for the same route path this method returns the configured
+     * navigation target for the path.
+     *
+     * @return an optional existing navigation target in the configuration which
+     *         caused the exception, or an empty optional if the exception is
+     *         not caused by a collision, not {@code null}
+     */
+    public Class<? extends Component> getConfiguredNavigationTarget() {
+        return navigationTarget;
+    }
+
+}

--- a/flow-server/src/main/java/com/vaadin/flow/server/InvalidRouteConfigurationException.java
+++ b/flow-server/src/main/java/com/vaadin/flow/server/InvalidRouteConfigurationException.java
@@ -24,7 +24,7 @@ public class InvalidRouteConfigurationException extends RuntimeException {
     /**
      * Constructs a new invalid route configuration runtime exception with the
      * specified detail message.
-     * 
+     *
      * @param message
      *            the detail message. The detail message is saved for later
      *            retrieval by the {@link #getMessage()} method.

--- a/flow-server/src/main/java/com/vaadin/flow/server/startup/RouteRegistryInitializer.java
+++ b/flow-server/src/main/java/com/vaadin/flow/server/startup/RouteRegistryInitializer.java
@@ -15,16 +15,19 @@
  */
 package com.vaadin.flow.server.startup;
 
+import java.util.Set;
+
 import javax.servlet.ServletContainerInitializer;
 import javax.servlet.ServletContext;
 import javax.servlet.ServletException;
 import javax.servlet.annotation.HandlesTypes;
-import java.util.Set;
 
+import com.googlecode.gentyref.GenericTypeReflector;
 import com.vaadin.flow.component.Component;
 import com.vaadin.flow.router.Route;
 import com.vaadin.flow.router.RouteAlias;
 import com.vaadin.flow.router.RouteConfiguration;
+import com.vaadin.flow.server.AmbiguousRouteConfigurationException;
 import com.vaadin.flow.server.InvalidRouteConfigurationException;
 
 /**
@@ -53,12 +56,8 @@ public class RouteRegistryInitializer extends AbstractRouteRegistryInitializer
 
             RouteConfiguration routeConfiguration = RouteConfiguration
                     .forRegistry(routeRegistry);
-            routeConfiguration.update(() -> {
-                routeConfiguration.getHandledRegistry().clean();
-                for (Class<? extends Component> navigationTarget : routes) {
-                    routeConfiguration.setAnnotatedRoute(navigationTarget);
-                }
-            });
+            routeConfiguration.update(
+                    () -> setAnnotatedRoutes(routeConfiguration, routes));
             routeRegistry.setPwaConfigurationClass(validatePwaClass(
                     routes.stream().map(clazz -> (Class<?>) clazz)));
         } catch (InvalidRouteConfigurationException irce) {
@@ -66,6 +65,37 @@ public class RouteRegistryInitializer extends AbstractRouteRegistryInitializer
                     "Exception while registering Routes on servlet startup",
                     irce);
         }
+    }
+
+    private void setAnnotatedRoutes(RouteConfiguration routeConfiguration,
+            Set<Class<? extends Component>> routes) {
+        routeConfiguration.getHandledRegistry().clean();
+        for (Class<? extends Component> navigationTarget : routes) {
+            try {
+                routeConfiguration.setAnnotatedRoute(navigationTarget);
+            } catch (AmbiguousRouteConfigurationException exception) {
+                if (!handleAmbiguousRoute(routeConfiguration,
+                        exception.getConfiguredNavigationTarget(),
+                        navigationTarget)) {
+                    throw exception;
+                }
+            }
+        }
+    }
+
+    private boolean handleAmbiguousRoute(RouteConfiguration routeConfiguration,
+            Class<? extends Component> configuredNavigationTarget,
+            Class<? extends Component> navigationTarget) {
+        if (GenericTypeReflector.isSuperType(navigationTarget,
+                configuredNavigationTarget)) {
+            return true;
+        } else if (GenericTypeReflector.isSuperType(configuredNavigationTarget,
+                navigationTarget)) {
+            routeConfiguration.removeRoute(configuredNavigationTarget);
+            routeConfiguration.setAnnotatedRoute(navigationTarget);
+            return true;
+        }
+        return false;
     }
 
 }

--- a/flow-server/src/main/java/com/vaadin/flow/server/startup/RouteTarget.java
+++ b/flow-server/src/main/java/com/vaadin/flow/server/startup/RouteTarget.java
@@ -28,6 +28,7 @@ import com.vaadin.flow.router.OptionalParameter;
 import com.vaadin.flow.router.ParameterDeserializer;
 import com.vaadin.flow.router.RouterLayout;
 import com.vaadin.flow.router.WildcardParameter;
+import com.vaadin.flow.server.AmbiguousRouteConfigurationException;
 import com.vaadin.flow.server.InvalidRouteConfigurationException;
 
 /**
@@ -55,10 +56,10 @@ public class RouteTarget implements Serializable {
      * Note! This will create a mutable RouteTarget by default.
      *
      * @param target
-     *         navigation target
+     *            navigation target
      * @throws InvalidRouteConfigurationException
-     *         exception for miss configured routes where navigation targets
-     *         can not be clearly selected
+     *             exception for miss configured routes where navigation targets
+     *             can not be clearly selected
      */
     public RouteTarget(Class<? extends Component> target) {
         this(target, true);
@@ -68,12 +69,12 @@ public class RouteTarget implements Serializable {
      * Create a new Route target holder with the given target registered.
      *
      * @param target
-     *         navigation target
+     *            navigation target
      * @param mutable
-     *         if this should be mutable
+     *            if this should be mutable
      * @throws InvalidRouteConfigurationException
-     *         exception for miss configured routes where navigation targets
-     *         can not be clearly selected
+     *             exception for miss configured routes where navigation targets
+     *             can not be clearly selected
      */
     public RouteTarget(Class<? extends Component> target, boolean mutable) {
         this.mutable = mutable;
@@ -87,28 +88,29 @@ public class RouteTarget implements Serializable {
      * add with the already existing navigation targets.
      *
      * @param target
-     *         navigation target to add
+     *            navigation target to add
      * @throws InvalidRouteConfigurationException
-     *         exception for miss configured routes where navigation targets
-     *         can not be clearly selected
+     *             exception for miss configured routes where navigation targets
+     *             can not be clearly selected
      */
     public void addRoute(Class<? extends Component> target) {
         throwIfImmutable();
         addTargetByType(target);
     }
 
-    private void addTargetByType(Class<? extends Component> target) {
+    private void addTargetByType(Class<? extends Component> target)
+            throws InvalidRouteConfigurationException {
         if (!HasUrlParameter.class.isAssignableFrom(target)
                 && !isAnnotatedParameter(target)) {
             validateNormalTarget(target);
             normal = target;
         } else {
-            if (ParameterDeserializer
-                    .isAnnotatedParameter(target, OptionalParameter.class)) {
+            if (ParameterDeserializer.isAnnotatedParameter(target,
+                    OptionalParameter.class)) {
                 validateOptionalParameter(target);
                 optionalParameter = target;
-            } else if (ParameterDeserializer
-                    .isAnnotatedParameter(target, WildcardParameter.class)) {
+            } else if (ParameterDeserializer.isAnnotatedParameter(target,
+                    WildcardParameter.class)) {
                 validateWildcard(target);
                 wildCardParameter = target;
             } else {
@@ -121,46 +123,49 @@ public class RouteTarget implements Serializable {
     private void validateParameter(Class<? extends Component> target)
             throws InvalidRouteConfigurationException {
         if (parameter != null) {
-            throw new InvalidRouteConfigurationException(String.format(
+            throw new AmbiguousRouteConfigurationException(String.format(
                     "Navigation targets must have unique routes, found navigation targets '%s' and '%s' with parameter have the same route.",
-                    parameter.getName(), target.getName()));
+                    parameter.getName(), target.getName()), parameter);
         }
     }
 
     private void validateWildcard(Class<? extends Component> target)
             throws InvalidRouteConfigurationException {
         if (wildCardParameter != null) {
-            throw new InvalidRouteConfigurationException(String.format(
+            throw new AmbiguousRouteConfigurationException(String.format(
                     "Navigation targets must have unique routes, found navigation targets '%s' and '%s' with wildcard parameter have the same route.",
-                    wildCardParameter.getName(), target.getName()));
+                    wildCardParameter.getName(), target.getName()),
+                    wildCardParameter);
         }
     }
 
     private void validateOptionalParameter(Class<? extends Component> target)
             throws InvalidRouteConfigurationException {
         if (normal != null) {
-            throw new InvalidRouteConfigurationException(String.format(
+            throw new AmbiguousRouteConfigurationException(String.format(
                     "Navigation targets '%s' and '%s' have the same path and '%s' has an OptionalParameter that will never be used as optional.",
-                    normal.getName(), target.getName(), target.getName()));
+                    normal.getName(), target.getName(), target.getName()),
+                    normal);
         } else if (optionalParameter != null) {
-            String message = String
-                    .format("Navigation targets must have unique routes, found navigation targets '%s' and '%s' with parameter have the same route.",
-                            optionalParameter.getName(), target.getName());
-            throw new InvalidRouteConfigurationException(message);
+            String message = String.format(
+                    "Navigation targets must have unique routes, found navigation targets '%s' and '%s' with parameter have the same route.",
+                    optionalParameter.getName(), target.getName());
+            throw new AmbiguousRouteConfigurationException(message,
+                    optionalParameter);
         }
     }
 
     private void validateNormalTarget(Class<? extends Component> target)
             throws InvalidRouteConfigurationException {
         if (normal != null) {
-            throw new InvalidRouteConfigurationException(String.format(
+            throw new AmbiguousRouteConfigurationException(String.format(
                     "Navigation targets must have unique routes, found navigation targets '%s' and '%s' with the same route.",
-                    normal.getName(), target.getName()));
+                    normal.getName(), target.getName()), normal);
         } else if (optionalParameter != null) {
-            throw new InvalidRouteConfigurationException(String.format(
+            throw new AmbiguousRouteConfigurationException(String.format(
                     "Navigation targets '%s' and '%s' have the same path and '%s' has an OptionalParameter that will never be used as optional.",
                     target.getName(), optionalParameter.getName(),
-                    optionalParameter.getName()));
+                    optionalParameter.getName()), optionalParameter);
         }
     }
 
@@ -168,7 +173,7 @@ public class RouteTarget implements Serializable {
      * Get route target for given segments.
      *
      * @param segments
-     *         route segments
+     *            route segments
      * @return navigation target corresponding to given segments
      */
     public Class<? extends Component> getTarget(List<String> segments) {
@@ -185,17 +190,17 @@ public class RouteTarget implements Serializable {
     }
 
     private boolean isAnnotatedParameter(Class<?> target) {
-        return ParameterDeserializer
-                .isAnnotatedParameter(target, OptionalParameter.class)
-                || ParameterDeserializer
-                .isAnnotatedParameter(target, WildcardParameter.class);
+        return ParameterDeserializer.isAnnotatedParameter(target,
+                OptionalParameter.class)
+                || ParameterDeserializer.isAnnotatedParameter(target,
+                        WildcardParameter.class);
     }
 
     /**
      * Create a copy of this RouteTarget.
      *
      * @param mutable
-     *         if created copy is mutable or not
+     *            if created copy is mutable or not
      * @return copy of this RouteTarget
      */
     public RouteTarget copy(boolean mutable) {
@@ -210,11 +215,11 @@ public class RouteTarget implements Serializable {
     }
 
     /**
-     * Remove target route from this RouteTarget.
-     * This will also clear the parent layout chain for the target.
+     * Remove target route from this RouteTarget. This will also clear the
+     * parent layout chain for the target.
      *
      * @param targetRoute
-     *         route to remove
+     *            route to remove
      */
     public void remove(Class<? extends Component> targetRoute) {
         throwIfImmutable();
@@ -235,7 +240,7 @@ public class RouteTarget implements Serializable {
      * Check if navigation target is present in current target.
      *
      * @param target
-     *         navigation target to check for
+     *            navigation target to check for
      * @return true if navigation target is found in some position
      */
     public boolean containsTarget(Class<? extends Component> target) {
@@ -284,9 +289,9 @@ public class RouteTarget implements Serializable {
      * IllegalArgumentException will be thrown.
      *
      * @param target
-     *         target to add chain for
+     *            target to add chain for
      * @param parents
-     *         parent layout chain
+     *            parent layout chain
      */
     public void setParentLayouts(Class<? extends Component> target,
             List<Class<? extends RouterLayout>> parents) {
@@ -304,7 +309,7 @@ public class RouteTarget implements Serializable {
      * Get the parent layout chain defined for the given target.
      *
      * @param target
-     *         target to get parent layout chain for
+     *            target to get parent layout chain for
      * @return parent layout chain
      */
     public List<Class<? extends RouterLayout>> getParentLayouts(

--- a/flow-server/src/test/java/com/vaadin/flow/router/RouteConfigurationTest.java
+++ b/flow-server/src/test/java/com/vaadin/flow/router/RouteConfigurationTest.java
@@ -1,15 +1,14 @@
 package com.vaadin.flow.router;
 
-import javax.servlet.ServletContext;
 import java.util.ArrayList;
 import java.util.Arrays;
 import java.util.Collections;
 import java.util.List;
 import java.util.Optional;
-import java.util.stream.Collectors;
-import java.util.stream.Stream;
+import java.util.concurrent.CountDownLatch;
 
-import net.jcip.annotations.NotThreadSafe;
+import javax.servlet.ServletContext;
+
 import org.junit.Assert;
 import org.junit.Before;
 import org.junit.Test;
@@ -27,6 +26,8 @@ import com.vaadin.flow.server.VaadinServlet;
 import com.vaadin.flow.server.VaadinServletService;
 import com.vaadin.flow.server.VaadinSession;
 import com.vaadin.flow.server.startup.ApplicationRouteRegistry;
+
+import net.jcip.annotations.NotThreadSafe;
 
 @NotThreadSafe
 public class RouteConfigurationTest {
@@ -61,7 +62,7 @@ public class RouteConfigurationTest {
      * Get registry by handing the session lock correctly.
      *
      * @param session
-     *         target vaadin session
+     *            target vaadin session
      * @return session route registry for session if exists or new.
      */
     private SessionRouteRegistry getRegistry(VaadinSession session) {
@@ -75,22 +76,47 @@ public class RouteConfigurationTest {
     }
 
     @Test
-    public void routeConfigurationUpdateLock_newConfigurationIsGottenOnlyAfterUnlock() {
+    public void routeConfigurationUpdateLock_configurationIsUpdatedOnlyAfterUnlock() {
+        CountDownLatch waitReaderThread = new CountDownLatch(1);
+        CountDownLatch waitUpdaterThread = new CountDownLatch(2);
+
         RouteConfiguration routeConfiguration = RouteConfiguration
                 .forRegistry(getRegistry(session));
 
+        Thread readerThread = new Thread() {
+            @Override
+            public void run() {
+                awaitCountDown(waitUpdaterThread);
+
+                Assert.assertTrue("Registry should still remain empty",
+                        routeConfiguration.getAvailableRoutes().isEmpty());
+
+                awaitCountDown(waitUpdaterThread);
+
+                Assert.assertTrue("Registry should still remain empty",
+                        routeConfiguration.getAvailableRoutes().isEmpty());
+
+                waitReaderThread.countDown();
+            }
+        };
+
+        readerThread.start();
+
         routeConfiguration.update(() -> {
-            routeConfiguration
-                    .setRoute("", MyRoute.class, Collections.emptyList());
+            routeConfiguration.setRoute("", MyRoute.class,
+                    Collections.emptyList());
 
-            Assert.assertTrue("Registry should still remain empty",
-                    routeConfiguration.getAvailableRoutes().isEmpty());
+            waitUpdaterThread.countDown();
 
-            routeConfiguration
-                    .setRoute("path", Secondary.class, Collections.emptyList());
+            routeConfiguration.setRoute("path", Secondary.class,
+                    Collections.emptyList());
 
-            Assert.assertTrue("Registry should still remain empty",
-                    routeConfiguration.getAvailableRoutes().isEmpty());
+            waitUpdaterThread.countDown();
+            try {
+                waitReaderThread.await();
+            } catch (InterruptedException e) {
+                Assert.fail();
+            }
         });
 
         Assert.assertEquals(
@@ -104,10 +130,10 @@ public class RouteConfigurationTest {
                 .forRegistry(getRegistry(session));
 
         routeConfiguration.update(() -> {
-            routeConfiguration
-                    .setRoute("", MyRoute.class, Collections.emptyList());
-            routeConfiguration
-                    .setRoute("path", Secondary.class, Collections.emptyList());
+            routeConfiguration.setRoute("", MyRoute.class,
+                    Collections.emptyList());
+            routeConfiguration.setRoute("path", Secondary.class,
+                    Collections.emptyList());
         });
 
         Assert.assertTrue("Registered 'MyRoute.class' should return true",
@@ -126,8 +152,8 @@ public class RouteConfigurationTest {
         routeConfiguration.update(() -> {
             routeConfiguration.setRoute("", MyRoute.class);
             routeConfiguration.setRoute("path", Secondary.class);
-            routeConfiguration
-                    .setRoute("parents", MiddleLayout.class, MainLayout.class);
+            routeConfiguration.setRoute("parents", MiddleLayout.class,
+                    MainLayout.class);
         });
 
         Assert.assertEquals(
@@ -169,8 +195,8 @@ public class RouteConfigurationTest {
                 "'url' with no parameters should not have returned a class",
                 urlRoute.isPresent());
 
-        urlRoute = routeConfiguration
-                .getRoute("url", Collections.singletonList("param"));
+        urlRoute = routeConfiguration.getRoute("url",
+                Collections.singletonList("param"));
         Assert.assertTrue("'url' with parameters should have returned a class",
                 urlRoute.isPresent());
         Assert.assertEquals("'url' registration should be Url", Url.class,
@@ -329,22 +355,23 @@ public class RouteConfigurationTest {
         routeConfiguration.update(() -> {
             routeConfiguration.getHandledRegistry().clean();
             Arrays.asList(MyRoute.class, MyInfo.class, MyPalace.class,
-                    MyModular.class).forEach(routeConfiguration::setAnnotatedRoute);
+                    MyModular.class)
+                    .forEach(routeConfiguration::setAnnotatedRoute);
         });
 
         Mockito.verify(registry).update(Mockito.any());
 
-        Mockito.verify(registry)
-                .setRoute("home", MyRoute.class, Collections.emptyList());
+        Mockito.verify(registry).setRoute("home", MyRoute.class,
+                Collections.emptyList());
 
-        Mockito.verify(registry)
-                .setRoute("info", MyInfo.class, Collections.emptyList());
+        Mockito.verify(registry).setRoute("info", MyInfo.class,
+                Collections.emptyList());
 
-        Mockito.verify(registry)
-                .setRoute("palace", MyPalace.class, Collections.emptyList());
+        Mockito.verify(registry).setRoute("palace", MyPalace.class,
+                Collections.emptyList());
 
-        Mockito.verify(registry)
-                .setRoute("modular", MyModular.class, Collections.emptyList());
+        Mockito.verify(registry).setRoute("modular", MyModular.class,
+                Collections.emptyList());
     }
 
     @Test
@@ -355,9 +382,8 @@ public class RouteConfigurationTest {
 
         routeConfiguration.setAnnotatedRoute(MyRouteWithAliases.class);
 
-        Mockito.verify(registry)
-                .setRoute("withAliases", MyRouteWithAliases.class,
-                        Collections.emptyList());
+        Mockito.verify(registry).setRoute("withAliases",
+                MyRouteWithAliases.class, Collections.emptyList());
         Mockito.verify(registry).setRoute("version", MyRouteWithAliases.class,
                 Collections.emptyList());
         Mockito.verify(registry).setRoute("person", MyRouteWithAliases.class,
@@ -388,11 +414,19 @@ public class RouteConfigurationTest {
         RouteConfiguration routeConfiguration = RouteConfiguration
                 .forRegistry(registry);
 
-        routeConfiguration
-                .setParentAnnotatedRoute("middle", MiddleLayout.class);
+        routeConfiguration.setParentAnnotatedRoute("middle",
+                MiddleLayout.class);
 
         Mockito.verify(registry).setRoute("middle", MiddleLayout.class,
                 Collections.singletonList(MainLayout.class));
+    }
+
+    private void awaitCountDown(CountDownLatch countDownLatch) {
+        try {
+            countDownLatch.await();
+        } catch (InterruptedException e) {
+            Assert.fail();
+        }
     }
 
     @Tag("div")

--- a/flow-server/src/test/java/com/vaadin/flow/router/SessionRouteRegistryTest.java
+++ b/flow-server/src/test/java/com/vaadin/flow/router/SessionRouteRegistryTest.java
@@ -1,18 +1,20 @@
 package com.vaadin.flow.router;
 
-import javax.servlet.ServletContext;
 import java.util.ArrayList;
 import java.util.Arrays;
 import java.util.Collections;
 import java.util.List;
 import java.util.Optional;
 import java.util.concurrent.Callable;
+import java.util.concurrent.CountDownLatch;
 import java.util.concurrent.ExecutionException;
 import java.util.concurrent.ExecutorService;
 import java.util.concurrent.Executors;
 import java.util.concurrent.Future;
 import java.util.stream.Collectors;
 import java.util.stream.IntStream;
+
+import javax.servlet.ServletContext;
 
 import org.junit.Assert;
 import org.junit.Before;
@@ -21,7 +23,6 @@ import org.mockito.Mockito;
 
 import com.vaadin.flow.component.Component;
 import com.vaadin.flow.component.Tag;
-import com.vaadin.flow.component.UI;
 import com.vaadin.flow.server.MockVaadinSession;
 import com.vaadin.flow.server.RouteRegistry;
 import com.vaadin.flow.server.SessionRouteRegistry;
@@ -59,7 +60,7 @@ public class SessionRouteRegistryTest {
      * Get registry by handing the session lock correctly.
      *
      * @param session
-     *         target vaadin session
+     *            target vaadin session
      * @return session route registry for session if exists or new.
      */
     private SessionRouteRegistry getRegistry(VaadinSession session) {
@@ -139,8 +140,8 @@ public class SessionRouteRegistryTest {
     public void sessionRegistryOverridesParentRegistryForGetTargetUrl_globalRouteStillAccessible() {
         registry.setRoute("MyRoute", MyRoute.class, Collections.emptyList());
         SessionRouteRegistry sessionRegistry = getRegistry(session);
-        sessionRegistry
-                .setRoute("alternate", MyRoute.class, Collections.emptyList());
+        sessionRegistry.setRoute("alternate", MyRoute.class,
+                Collections.emptyList());
 
         Assert.assertEquals("Expected session registry route to be returned",
                 "alternate", sessionRegistry.getTargetUrl(MyRoute.class).get());
@@ -155,8 +156,8 @@ public class SessionRouteRegistryTest {
     public void sessionRegistryOverridesParentRegistryWithOwnClass_globalRouteReturnedAfterClassRemoval() {
         registry.setRoute("MyRoute", MyRoute.class, Collections.emptyList());
         SessionRouteRegistry sessionRegistry = getRegistry(session);
-        sessionRegistry
-                .setRoute("MyRoute", Secondary.class, Collections.emptyList());
+        sessionRegistry.setRoute("MyRoute", Secondary.class,
+                Collections.emptyList());
 
         Assert.assertEquals(
                 "Route 'MyRoute' should return Secondary as registered to SessionRegistry.",
@@ -291,24 +292,24 @@ public class SessionRouteRegistryTest {
                 sessionRegistry.getTargetUrl(MyRouteWithAliases.class)
                         .isPresent());
 
-        // Either or is expected as the new default as first match is picked from the map
+        // Either or is expected as the new default as first match is picked
+        // from the map
         Assert.assertTrue(
                 "Route didn't return a url matching either of the expected aliases.",
-                Arrays.asList("info", "version").contains(
-                        sessionRegistry.getTargetUrl(MyRouteWithAliases.class)
-                                .get()));
+                Arrays.asList("info", "version").contains(sessionRegistry
+                        .getTargetUrl(MyRouteWithAliases.class).get()));
     }
 
     @Test
     public void manuallyRegisteredAliases_RouteDataIsReturnedCorrectly() {
 
         SessionRouteRegistry sessionRegistry = getRegistry(session);
-        sessionRegistry
-                .setRoute("main", Secondary.class, Collections.emptyList());
-        sessionRegistry
-                .setRoute("Alias1", Secondary.class, Collections.emptyList());
-        sessionRegistry
-                .setRoute("Alias2", Secondary.class, Collections.emptyList());
+        sessionRegistry.setRoute("main", Secondary.class,
+                Collections.emptyList());
+        sessionRegistry.setRoute("Alias1", Secondary.class,
+                Collections.emptyList());
+        sessionRegistry.setRoute("Alias2", Secondary.class,
+                Collections.emptyList());
 
         List<RouteData> registeredRoutes = sessionRegistry
                 .getRegisteredRoutes();
@@ -378,8 +379,8 @@ public class SessionRouteRegistryTest {
 
         SessionRouteRegistry sessionRegistry = getRegistry(session);
 
-        sessionRegistry
-                .setRoute("MyRoute", Secondary.class, Collections.emptyList());
+        sessionRegistry.setRoute("MyRoute", Secondary.class,
+                Collections.emptyList());
 
         Assert.assertTrue("Registry didn't contain routes.",
                 !sessionRegistry.getRegisteredRoutes().isEmpty());
@@ -439,9 +440,9 @@ public class SessionRouteRegistryTest {
         Assert.assertEquals(
                 "Expected 4 route already exists exceptions due to route target validation",
                 THREADS - 1, exceptions.size());
-        String expected = String
-                .format("Navigation targets must have unique routes, found navigation targets '%s' and '%s' with the same route.",
-                        MyRoute.class.getName(), MyRoute.class.getName());
+        String expected = String.format(
+                "Navigation targets must have unique routes, found navigation targets '%s' and '%s' with the same route.",
+                MyRoute.class.getName(), MyRoute.class.getName());
         for (String exception : exceptions) {
             Assert.assertEquals(expected, exception);
         }
@@ -491,9 +492,9 @@ public class SessionRouteRegistryTest {
         Assert.assertEquals(
                 "Expected 4 route already exists exceptions due to route target validation",
                 THREADS - 1, exceptions.size());
-        String expected = String
-                .format("Navigation targets must have unique routes, found navigation targets '%s' and '%s' with the same route.",
-                        MyRoute.class.getName(), MyRoute.class.getName());
+        String expected = String.format(
+                "Navigation targets must have unique routes, found navigation targets '%s' and '%s' with the same route.",
+                MyRoute.class.getName(), MyRoute.class.getName());
         for (String exception : exceptions) {
             Assert.assertEquals(expected, exception);
         }
@@ -570,10 +571,10 @@ public class SessionRouteRegistryTest {
     public void updateAndRemoveFromMultipleThreads_endResultAsExpected()
             throws InterruptedException, ExecutionException {
 
-        getRegistry(session)
-                .setRoute("home", MyRoute.class, Collections.emptyList());
-        getRegistry(session)
-                .setRoute("info", MyRoute.class, Collections.emptyList());
+        getRegistry(session).setRoute("home", MyRoute.class,
+                Collections.emptyList());
+        getRegistry(session).setRoute("info", MyRoute.class,
+                Collections.emptyList());
 
         List<Callable<Result>> callables = new ArrayList<>();
         callables.add(() -> {
@@ -676,19 +677,40 @@ public class SessionRouteRegistryTest {
     }
 
     @Test
-    public void lockingConfiguration_newConfigurationIsGottenOnlyAfterUnlock() {
+    public void lockingConfiguration_configurationIsUpdatedOnlyAfterUnlockk() {
+        CountDownLatch waitReaderThread = new CountDownLatch(1);
+        CountDownLatch waitUpdaterThread = new CountDownLatch(2);
+
         SessionRouteRegistry registry = getRegistry(session);
+
+        Thread readerThread = new Thread() {
+            @Override
+            public void run() {
+                awaitCountDown(waitUpdaterThread);
+
+                Assert.assertTrue("Registry should still remain empty",
+                        getRegistry(session).getRegisteredRoutes().isEmpty());
+
+                awaitCountDown(waitUpdaterThread);
+
+                Assert.assertTrue("Registry should still remain empty",
+                        getRegistry(session).getRegisteredRoutes().isEmpty());
+
+                waitReaderThread.countDown();
+            }
+        };
+
+        readerThread.start();
 
         registry.update(() -> {
             registry.setRoute("", MyRoute.class, Collections.emptyList());
 
-            Assert.assertTrue("Registry should still remain empty",
-                    getRegistry(session).getRegisteredRoutes().isEmpty());
+            waitUpdaterThread.countDown();
 
             registry.setRoute("path", Secondary.class, Collections.emptyList());
 
-            Assert.assertTrue("Registry should still remain empty",
-                    getRegistry(session).getRegisteredRoutes().isEmpty());
+            waitUpdaterThread.countDown();
+            awaitCountDown(waitReaderThread);
         });
 
         Assert.assertEquals(
@@ -807,8 +829,8 @@ public class SessionRouteRegistryTest {
         });
 
         sessionRegistry.update(() -> {
-            sessionRegistry
-                    .setRoute("main", Secondary.class, Collections.emptyList());
+            sessionRegistry.setRoute("main", Secondary.class,
+                    Collections.emptyList());
             sessionRegistry.setRoute("Alias1", Secondary.class,
                     Collections.emptyList());
             sessionRegistry.setRoute("Alias2", Secondary.class,
@@ -838,8 +860,8 @@ public class SessionRouteRegistryTest {
         List<RoutesChangedEvent> events = new ArrayList<>();
 
         sessionRegistry.update(() -> {
-            sessionRegistry
-                    .setRoute("main", Secondary.class, Collections.emptyList());
+            sessionRegistry.setRoute("main", Secondary.class,
+                    Collections.emptyList());
             sessionRegistry.setRoute("Alias1", Secondary.class,
                     Collections.emptyList());
             sessionRegistry.setRoute("Alias2", Secondary.class,
@@ -883,26 +905,40 @@ public class SessionRouteRegistryTest {
 
         registry.setRoute("main", MyRoute.class, Collections.emptyList());
         sessionRegistry.update(() -> {
-            sessionRegistry
-                    .setRoute("main", Secondary.class, Collections.emptyList());
+            sessionRegistry.setRoute("main", Secondary.class,
+                    Collections.emptyList());
             sessionRegistry.setRoute("Alias1", Secondary.class,
                     Collections.emptyList());
             sessionRegistry.setRoute("Alias2", Secondary.class,
                     Collections.emptyList());
         });
 
-        Assert.assertEquals("One event for both registries should have been fired.", 2, events.size());
+        Assert.assertEquals(
+                "One event for both registries should have been fired.", 2,
+                events.size());
 
         registration.remove();
 
         sessionRegistry.removeRoute("main");
 
-        Assert.assertEquals("No new event should have been received for session scope", 2, events.size());
+        Assert.assertEquals(
+                "No new event should have been received for session scope", 2,
+                events.size());
 
         registry.removeRoute("main");
 
-        Assert.assertEquals("No new event should have been received for application scope", 2, events.size());
+        Assert.assertEquals(
+                "No new event should have been received for application scope",
+                2, events.size());
 
+    }
+
+    private void awaitCountDown(CountDownLatch countDownLatch) {
+        try {
+            countDownLatch.await();
+        } catch (InterruptedException e) {
+            Assert.fail();
+        }
     }
 
     @Tag("div")

--- a/flow-server/src/test/java/com/vaadin/flow/router/internal/AbstractRouteRegistryTest.java
+++ b/flow-server/src/test/java/com/vaadin/flow/router/internal/AbstractRouteRegistryTest.java
@@ -5,6 +5,11 @@ import java.util.Collections;
 import java.util.List;
 import java.util.Objects;
 import java.util.Optional;
+import java.util.concurrent.CountDownLatch;
+
+import org.junit.Assert;
+import org.junit.Before;
+import org.junit.Test;
 
 import com.vaadin.flow.component.Component;
 import com.vaadin.flow.component.Tag;
@@ -13,9 +18,6 @@ import com.vaadin.flow.router.RouteBaseData;
 import com.vaadin.flow.router.RouterLayout;
 import com.vaadin.flow.router.RoutesChangedEvent;
 import com.vaadin.flow.shared.Registration;
-import org.junit.Assert;
-import org.junit.Before;
-import org.junit.Test;
 
 public class AbstractRouteRegistryTest {
 
@@ -44,18 +46,32 @@ public class AbstractRouteRegistryTest {
     }
 
     @Test
-    public void lockingConfiguration_newConfigurationIsGottenOnlyAfterUnlock() {
+    public void lockingConfiguration_configurationIsUpdatedOnlyAfterUnlock() {
+        CountDownLatch waitReaderThread = new CountDownLatch(1);
+        CountDownLatch waitUpdaterThread = new CountDownLatch(2);
+
+        Thread readerThread = new Thread() {
+            @Override
+            public void run() {
+                awaitCountDown(waitUpdaterThread);
+
+                Assert.assertTrue("Registry should still remain empty",
+                        registry.getRegisteredRoutes().isEmpty());
+
+                awaitCountDown(waitUpdaterThread);
+
+                Assert.assertTrue("Registry should still remain empty",
+                        registry.getRegisteredRoutes().isEmpty());
+
+                waitReaderThread.countDown();
+            }
+        };
+
+        readerThread.start();
 
         registry.update(() -> {
             registry.setRoute("", MyRoute.class, Collections.emptyList());
-
-            Assert.assertTrue("Registry should still remain empty",
-                    registry.getRegisteredRoutes().isEmpty());
-
             registry.setRoute("path", Secondary.class, Collections.emptyList());
-
-            Assert.assertTrue("Registry should still remain empty",
-                    registry.getRegisteredRoutes().isEmpty());
         });
 
         Assert.assertEquals(
@@ -244,6 +260,14 @@ public class AbstractRouteRegistryTest {
         registry.setRoute("away", MyRoute.class, Collections.emptyList());
 
         Assert.assertEquals("No new event should have fired", 1, events.size());
+    }
+
+    private void awaitCountDown(CountDownLatch countDownLatch) {
+        try {
+            countDownLatch.await();
+        } catch (InterruptedException e) {
+            Assert.fail();
+        }
     }
 
     @Tag("div")

--- a/flow-server/src/test/java/com/vaadin/flow/server/startup/ApplicationRouteRegistryTest.java
+++ b/flow-server/src/test/java/com/vaadin/flow/server/startup/ApplicationRouteRegistryTest.java
@@ -1,21 +1,24 @@
 package com.vaadin.flow.server.startup;
 
-import javax.servlet.ServletContext;
 import java.util.ArrayList;
 import java.util.Collections;
 import java.util.List;
 import java.util.concurrent.Callable;
+import java.util.concurrent.CountDownLatch;
 import java.util.concurrent.ExecutionException;
 import java.util.concurrent.ExecutorService;
 import java.util.concurrent.Executors;
 import java.util.concurrent.Future;
 
-import com.vaadin.flow.router.RouteBaseData;
-import com.vaadin.flow.server.RouteRegistry;
+import javax.servlet.ServletContext;
+
 import org.junit.Assert;
 import org.junit.Before;
 import org.junit.Test;
 import org.mockito.Mockito;
+
+import com.vaadin.flow.router.RouteBaseData;
+import com.vaadin.flow.server.RouteRegistry;
 
 /**
  * Tests for {@link ApplicationRouteRegistry} instance inside OSGi container.
@@ -179,18 +182,40 @@ public class ApplicationRouteRegistryTest extends RouteRegistryTestBase {
 
     @Test
     public void lockingConfiguration_newConfigurationIsGottenOnlyAfterUnlock() {
+        CountDownLatch waitReaderThread = new CountDownLatch(1);
+        CountDownLatch waitUpdaterThread = new CountDownLatch(2);
+
+        Thread readerThread = new Thread() {
+            @Override
+            public void run() {
+                awaitCountDown(waitUpdaterThread);
+
+                Assert.assertTrue("Registry should still remain empty",
+                        getTestedRegistry().getRegisteredRoutes().isEmpty());
+
+                awaitCountDown(waitUpdaterThread);
+
+                Assert.assertTrue("Registry should still remain empty",
+                        getTestedRegistry().getRegisteredRoutes().isEmpty());
+
+                waitReaderThread.countDown();
+            }
+        };
+
+        readerThread.start();
+
         getTestedRegistry().update(() -> {
             getTestedRegistry().setRoute("", MyRoute.class,
                     Collections.emptyList());
 
-            Assert.assertTrue("Registry should still remain empty",
-                    getTestedRegistry().getRegisteredRoutes().isEmpty());
+            waitUpdaterThread.countDown();
 
             getTestedRegistry().setRoute("path", Secondary.class,
                     Collections.emptyList());
 
-            Assert.assertTrue("Registry should still remain empty",
-                    getTestedRegistry().getRegisteredRoutes().isEmpty());
+            waitUpdaterThread.countDown();
+            awaitCountDown(waitReaderThread);
+
         });
 
         Assert.assertEquals(
@@ -336,6 +361,14 @@ public class ApplicationRouteRegistryTest extends RouteRegistryTestBase {
     @Override
     protected RouteRegistry getTestedRegistry() {
         return registry;
+    }
+
+    private void awaitCountDown(CountDownLatch countDownLatch) {
+        try {
+            countDownLatch.await();
+        } catch (InterruptedException e) {
+            Assert.fail();
+        }
     }
 
 }

--- a/flow-server/src/test/java/com/vaadin/flow/server/startup/RouteRegistryInitializerTest.java
+++ b/flow-server/src/test/java/com/vaadin/flow/server/startup/RouteRegistryInitializerTest.java
@@ -15,15 +15,17 @@
  */
 package com.vaadin.flow.server.startup;
 
-import javax.servlet.ServletContext;
-import javax.servlet.ServletException;
 import java.util.Arrays;
 import java.util.Collections;
+import java.util.LinkedHashSet;
 import java.util.List;
 import java.util.Optional;
 import java.util.Set;
 import java.util.stream.Collectors;
 import java.util.stream.Stream;
+
+import javax.servlet.ServletContext;
+import javax.servlet.ServletException;
 
 import org.junit.Assert;
 import org.junit.Before;
@@ -127,7 +129,8 @@ public class RouteRegistryInitializerTest {
             throws ServletException {
         routeRegistryInitializer.onStartup(
                 Stream.of(NavigationTargetBar.class, NavigationTargetBar2.class)
-                        .collect(Collectors.toSet()), servletContext);
+                        .collect(Collectors.toSet()),
+                servletContext);
     }
 
     @Test
@@ -890,8 +893,8 @@ public class RouteRegistryInitializerTest {
                 Parent.class.getName(),
                 RootConfiguratorWithParent.class.getName()));
 
-        routeRegistryInitializer.onStartup(
-                Stream.of(RootConfiguratorWithParent.class)
+        routeRegistryInitializer
+                .onStartup(Stream.of(RootConfiguratorWithParent.class)
                         .collect(Collectors.toSet()), servletContext);
     }
 
@@ -1439,4 +1442,111 @@ public class RouteRegistryInitializerTest {
                 errorTargetEntry.getNavigationTarget());
     }
 
+    @Test
+    public void registerClassesWithSameRoute_abstractClass_subclass_subclassIsRegistered()
+            throws ServletException {
+        LinkedHashSet<Class<?>> classes = new LinkedHashSet<>();
+        classes.add(AbstractRouteTarget.class);
+        classes.add(BaseRouteTarget.class);
+        routeRegistryInitializer.onStartup(classes, servletContext);
+
+        Assert.assertEquals(BaseRouteTarget.class,
+                registry.getNavigationTarget("foo").get());
+    }
+
+    @Test
+    public void registerClassesWithSameRoute_class_abstractSuperClass_subclassIsRegistered()
+            throws ServletException {
+        LinkedHashSet<Class<?>> classes = new LinkedHashSet<>();
+        classes.add(BaseRouteTarget.class);
+        classes.add(AbstractRouteTarget.class);
+        routeRegistryInitializer.onStartup(classes, servletContext);
+
+        Assert.assertEquals(BaseRouteTarget.class,
+                registry.getNavigationTarget("foo").get());
+    }
+
+    @Test
+    public void registerClassesWithSameRoute_class_subclass_subclassIsRegistered()
+            throws ServletException {
+        LinkedHashSet<Class<?>> classes = new LinkedHashSet<>();
+        classes.add(BaseRouteTarget.class);
+        classes.add(SuperRouteTarget.class);
+        routeRegistryInitializer.onStartup(classes, servletContext);
+
+        Assert.assertEquals(SuperRouteTarget.class,
+                registry.getNavigationTarget("foo").get());
+    }
+
+    @Test
+    public void registerClassesWithSameRoute_class_superClass_subclassIsRegistered()
+            throws ServletException {
+        LinkedHashSet<Class<?>> classes = new LinkedHashSet<>();
+        classes.add(SuperRouteTarget.class);
+        classes.add(BaseRouteTarget.class);
+        routeRegistryInitializer.onStartup(classes, servletContext);
+
+        Assert.assertEquals(SuperRouteTarget.class,
+                registry.getNavigationTarget("foo").get());
+    }
+
+    @Test(expected = ServletException.class)
+    public void registerClassesWithSameRoute_absatrctClass_unrelatedClass_throws()
+            throws ServletException {
+        LinkedHashSet<Class<?>> classes = new LinkedHashSet<>();
+        classes.add(AbstractRouteTarget.class);
+        classes.add(OtherRouteTarget.class);
+        routeRegistryInitializer.onStartup(classes, servletContext);
+    }
+
+    @Test(expected = ServletException.class)
+    public void registerClassesWithSameRoute_unrelatedClass_abstractClass_throws()
+            throws ServletException {
+        LinkedHashSet<Class<?>> classes = new LinkedHashSet<>();
+        classes.add(OtherRouteTarget.class);
+        classes.add(AbstractRouteTarget.class);
+        routeRegistryInitializer.onStartup(classes, servletContext);
+    }
+
+    @Test(expected = ServletException.class)
+    public void registerClassesWithSameRoute_class_unrelatedClass_throws()
+            throws ServletException {
+        LinkedHashSet<Class<?>> classes = new LinkedHashSet<>();
+        classes.add(BaseRouteTarget.class);
+        classes.add(OtherRouteTarget.class);
+        routeRegistryInitializer.onStartup(classes, servletContext);
+    }
+
+    @Test(expected = ServletException.class)
+    public void registerClassesWithSameRoute_unrelatedClass_class_throws()
+            throws ServletException {
+        LinkedHashSet<Class<?>> classes = new LinkedHashSet<>();
+        classes.add(OtherRouteTarget.class);
+        classes.add(BaseRouteTarget.class);
+        routeRegistryInitializer.onStartup(classes, servletContext);
+    }
+
+    @Tag(Tag.DIV)
+    @Route("foo")
+    private abstract static class AbstractRouteTarget extends Component {
+
+    }
+
+    @Tag(Tag.DIV)
+    @Route("foo")
+    private static class BaseRouteTarget extends AbstractRouteTarget {
+
+    }
+
+    @Tag(Tag.DIV)
+    @Route("foo")
+    private static class SuperRouteTarget extends BaseRouteTarget {
+
+    }
+
+    @Tag(Tag.DIV)
+    @Route("foo")
+    private static class OtherRouteTarget extends Component {
+
+    }
 }

